scalafix-core/src/main/scala/scalafix/rewrite/Rewrite.scala

```diff
@@ -1,31 +1,59 @@
-package scalafix.rewrite
+package scalafix
+package rewrite
 
 import scala.collection.immutable.Seq
-import scala.meta._
-import scalafix.util.Patch
 import scala.collection.immutable.Seq
+import scala.meta._
 import scalafix.config.ReaderUtil
 
-/** A rewrite is a named RewriteCtx[A] => Seq[Patch] function.
-  * @tparam A Required api in [[ScalafixMirror]]. Example values:
-  *           [[ScalafixMirror]] for scalafix-nsc,
-  *           [[scala.meta.Mirror]] when using scalahost or
-  *           [[Any]] for syntactic rewrites.
-  */
-abstract class Rewrite[-A](implicit sourceName: sourcecode.Name) {
+import metaconfig.ConfDecoder
+import sourcecode.Name
+
+/** A rewrite is a named RewriteCtx => Patch function. */
+abstract class Rewrite(implicit sourceName: Name) { self =>
+
   def name: String = sourceName.value
   override def toString: String = name
-  def rewrite[B <: A](ctx: RewriteCtx[B]): Seq[Patch]
+  def rewrite(ctx: RewriteCtx): Patch
+  def andThen(other: Rewrite): Rewrite = Rewrite.merge(this, other)
+  private[scalafix] def wrappedRewrite(ctx: RewriteCtx): Patch =
+    patch.InCtx(rewrite(ctx), ctx, None)
+}
 
-  def andThen[B <: A](other: Rewrite[B]): Rewrite[B] =
-    Rewrite(ctx => this.rewrite(ctx) ++ other.rewrite(ctx))
+abstract class SemanticRewrite(mirror: Mirror)(implicit name: Name)
+    extends Rewrite {
+  implicit val ImplicitMirror: Mirror = mirror
+  private[scalafix] override def wrappedRewrite(ctx: RewriteCtx): Patch =
+    patch.InCtx(rewrite(ctx), ctx, Some(mirror))
 }
 
 object Rewrite {
-  def withName[A](name: String)(f: RewriteCtx[A] => Seq[Patch]): Rewrite[A] =
-    apply(f)(sourcecode.Name(name))
-  def apply[T](f: RewriteCtx[T] => Seq[Patch])(
-      implicit name: sourcecode.Name): Rewrite[T] = new Rewrite[T]() {
-    override def rewrite[B <: T](ctx: RewriteCtx[B]): Seq[Patch] = f(ctx)
+  val syntaxRewriteConfDecoder = config.rewriteConfDecoder(None)
+  def empty: Rewrite = syntactic(_ => Patch.empty)
+  def combine(rewrites: Seq[Rewrite], mirror: Option[Mirror]): Rewrite =
+    rewrites.foldLeft(mirror.fold(empty)(emptySemantic))(_ andThen _)
+  def emptySemantic(mirror: Mirror): Rewrite =
+    semantic(x => y => Patch.empty)(Name("empty"))(mirror)
```

Comment:
Good question! I thought this was not necessary either so I removed it in one of the commits only to find out this is actually necessary to for example run the rewrites from .scalafix.conf `patches.renameSymbol=s/from/to`. In these cases we need an `InCtx(Some(mirror))` for symbols but the patch can be empty. 

I think there's a way to get rid of this one. I've added a comment explaining why it's here and a propsal on how to get rid of it eventually.